roseus/euslisp/roseus-utils.l

```diff
@@ -55,6 +55,68 @@
     ;; FIXME:: add coordinates of camera
     (make-camera-from-ros-camera-info-aux pwidth pheight p nil)))
 
+;; Image
+(unless (fboundp 'version-string-p)
+(defun version-string-p (str)
+  "Return T if STR is string of version defined below, return NIL otherwise.
+VERSION := NUMBER (DOT NUMBER)*
+NUMBER := (0|1|2|3|4|5|6|7|8|9)*"
+  (null (find-if-not #'(lambda (x)
+                         (and (integerp x) (not (minusp x))))
+                     (read-from-string (format nil "(~A)" (substitute #\Space #\. str)))))))
+(unless (fboundp 'version-string-to-list)
+(defun version-string-to-list (str)
+"Return list of version numbers of version string STR"
+  (unless (version-string-p str)
+    (error "version string must be number or dot: ~A~%" str))
+  (read-from-string (format nil "(~A)" (substitute #\Space #\. str)))))
+(unless (fboundp 'version-list-not-zero)
+(defun version-list-not-zero (lst)
+  "Return the first non-zero element of LST, which is a list of integers"
+  (while (and lst (zerop (car lst)))
+    (setq lst (cdr lst)))
+  (if lst (car lst) 0)))
+(unless (fboundp 'version<)
+(defun version< (ver1 ver2)
+  "Return T if version VER1 is lower (older) than VER2, return NIL otherwise"
+  (let ((l1 (version-string-to-list ver1))
+        (l2 (version-string-to-list ver2)))
+    (while (and l1 l2 (= (car l1) (car l2)))
+      (setq l1 (cdr l1) l2 (cdr l2)))
+    (cond
+     ((and l1 l2) (< (car l1) (car l2)))
+     ((and (null l1) (null l2)) nil)
+     (l1 (< (version-list-not-zero l1) 0))
+     (t (< 0 (version-list-not-zero l2)))))))
+
+;; this function is valid only on 1.0.12 and later
+(unless (version< (car (last lisp-implementation-version)) "1.0.11")
```

Comment:
instead of using version number, how about
```
(assoc :create-viewer (send camera-model :methods))
nil
```

--
◉ Kei Okada


On Thu, Nov 12, 2015 at 12:58 PM, Furushchev <notifications@github.com>
wrote:

> In roseus/euslisp/roseus-utils.l
> <https://github.com/jsk-ros-pkg/jsk_roseus/pull/393#discussion_r44619536>:
>
> > +  (if lst (car lst) 0)))
> > +(unless (fboundp 'version<)
> > +(defun version< (ver1 ver2)
> > +  "Return T if version VER1 is lower (older) than VER2, return NIL otherwise"
> > +  (let ((l1 (version-string-to-list ver1))
> > +        (l2 (version-string-to-list ver2)))
> > +    (while (and l1 l2 (= (car l1) (car l2)))
> > +      (setq l1 (cdr l1) l2 (cdr l2)))
> > +    (cond
> > +     ((and l1 l2) (< (car l1) (car l2)))
> > +     ((and (null l1) (null l2)) nil)
> > +     (l1 (< (version-list-not-zero l1) 0))
> > +     (t (< 0 (version-list-not-zero l2)))))))
> > +
> > +;; this function is valid only on 1.0.12 and later
> > +(unless (version< (car (last lisp-implementation-version)) "1.0.11")
>
> In
> https://github.com/jsk-ros-pkg/jsk_roseus/pull/393/files#diff-e5905e35fa4b963e4b009823daedcc76R121
> , loads demo file which is only available on 1.0.11.
> and in this file,
>
> https://github.com/euslisp/jskeus/blob/master/irteus/demo/sample-camera-model.l#L39
> :create-viewer method is available on 1.0.11.
> And I could not find to :get-image correctly without :create-viewer.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsk-ros-pkg/jsk_roseus/pull/393/files#r44619536>.
>